pkg/scheduler/actions/reclaim/reclaim.go

```diff
@@ -91,148 +91,142 @@ func (ra *Action) Execute(ssn *framework.Session) {
-		if !ssn.Preemptive(queue, task) {
-			klog.V(3).Infof("Queue <%s> can not reclaim by preempt others when considering task <%s> , ignore it.", queue.Name, task.Name)
-			continue
-		}
-
-		if err := ssn.PrePredicateFn(task); err != nil {
-			klog.V(3).Infof("PrePredicate for task %s/%s failed for: %v", task.Namespace, task.Name, err)
-			continue
-		}
-
-		assigned := false
-		// we should filter out those nodes that are UnschedulableAndUnresolvable status got in allocate action
-		totalNodes := ssn.FilterOutUnschedulableAndUnresolvableNodesForTask(task)
-		for _, n := range totalNodes {
-			// When filtering candidate nodes, need to consider the node statusSets instead of the err information.
-			// refer to kube-scheduler preemption code: https://github.com/kubernetes/kubernetes/blob/9d87fa215d9e8020abdc17132d1252536cd752d2/pkg/scheduler/framework/preemption/preemption.go#L422
-			if err := ssn.PredicateForPreemptAction(task, n); err != nil {
-				klog.V(4).Infof("Reclaim predicate for task %s/%s on node %s return error %v ", task.Namespace, task.Name, n.Name, err)
-				continue
+		for {
+			// Pick the starving jobs in this queue.
+			jobsQ, found := preemptorsMap[queue.UID]
+			if !found || jobsQ.Empty() {
+				klog.V(4).Infof("No preemptors in Queue <%s>, break.", queue.Name)
+				break
 			}
+			job := jobsQ.Pop().(*api.JobInfo)
+
+			var assignedInJob bool
+			for {
+				// Pick up all its candidate tasks.
+				tasksQ, ok := preemptorTasks[job.UID]
+				if !ok || tasksQ.Empty() || !ssn.JobStarving(job) {
+					klog.V(3).Infof("No preemptor task in job <%s/%s>.",
+						job.Namespace, job.Name)
+					break
+				}
 
-			klog.V(3).Infof("Considering Task <%s/%s> on Node <%s>.", task.Namespace, task.Name, n.Name)
+				klog.V(3).Infof("Try reclaim for %d tasks of job <%s/%s>", tasksQ.Len(), job.Namespace, job.Name)
 
-			var reclaimees []*api.TaskInfo
-			for _, task := range n.Tasks {
-				// Ignore non running task.
-				if task.Status != api.Running {
+				task := tasksQ.Pop().(*api.TaskInfo)
+
+				if task.Pod.Spec.PreemptionPolicy != nil && *task.Pod.Spec.PreemptionPolicy == v1.PreemptNever {
+					klog.V(3).Infof("Task %s/%s cannot preempt (policy Never)", task.Namespace, task.Name)
 					continue
 				}
-				if !task.Preemptable {
+
+				if !ssn.Preemptive(queue, task) {
+					klog.V(3).Infof("Queue <%s> cannot reclaim for task <%s>, skip", queue.Name, task.Name)
 					continue
 				}
 
-				if j, found := ssn.Jobs[task.Job]; !found {
+				if err := ssn.PrePredicateFn(task); err != nil {
+					klog.V(3).Infof("PrePredicate failed for task %s/%s: %v", task.Namespace, task.Name, err)
 					continue
-				} else if j.Queue != job.Queue {
-					q := ssn.Queues[j.Queue]
-					if !q.Reclaimable() {
-						continue
-					}
-					// Clone task to avoid modify Task's status on node.
-					reclaimees = append(reclaimees, task.Clone())
 				}
+
+				assignedInJob = assignedInJob || ra.reclaimForTask(ssn, task, job)
```

Comment:
I don't understand what does `assignedInJob` mean and what is this variable used for, could you clarify? Why would us push back the job into the jobsQ queue only `assignedInJob` is true?